crates/rune/src/languageserver/state.rs

```diff
@@ -250,6 +250,34 @@ impl<'a> State<'a> {
         Some(results)
     }
 
+    pub(super) fn format(&mut self, uri: &Url) -> Option<TextEdit> {
+        let sources = &mut self.workspace.sources;
+        tracing::trace!(uri = ?uri, uri_exists = sources.get(uri).is_some());
+        let workspace_source = sources.get_mut(uri)?;
+
+        // Not sure how to get source of `workspace_source`
+        let source = crate::source::Source::new(uri, workspace_source.content.to_string());
+        let formatted = crate::fmt::layout_source(&source).ok()?;
```

Comment:
The signature of layout_source doesn't make much sense forcing you to roundtrip, it could be simpler:
<https://github.com/rune-rs/rune/blob/96686814ad7626f99265856b6fad9c959d98bec3/crates/rune/src/fmt.rs#L22>.

Reply:
what exactly do you mean by that?

Reply:
Only thing I see would be to return `String` instead of `Vec<u8>`?

Reply:
I was thinking about changing the argument to be `source: &str`, I don't think the `Source` type is doing anything useful.

Since you didn't add it, I won't force you to do it. Only if you want to and no other issues pop up as a result.